iquip/apps/builder.py

```diff
@@ -1,32 +1,212 @@
+    QListWidgetItem, QPushButton, QVBoxLayout, QWidget
 )
 
 import qiwis
 from iquip.protocols import ExperimentInfo
 
+class _BaseEntry(QWidget):
+    """Base class for all argument entries.
+
+    This is a wrapper of each entry.
+    In each subclass, value() must be implemented to return the selected value.
+
+    Attributes:
+        name: The argument name.
+        argInfo: The dictionary with the argument options.
+        nameLabel: The label for showing the argument name.
+    """
+
+    def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget] = None):
+        """Extended.
+        
+        Args:
+            name, argInfo: See the attributes section in _BaseEntry.
+        """
+        super().__init__(parent=parent)
+        self.name = name
+        self.argInfo = argInfo
+        # widgets
+        self.nameLabel = QLabel(name, self)
+        # layout
+        self.layout = QHBoxLayout(self)
+        self.layout.addWidget(self.nameLabel)
+        self.setLayout(self.layout)
+
+    def value(self) -> Any:
+        """Returns the entered or selected value.
+        
+        This must be overridden in the subclass.
+        """
+        raise NotImplementedError
+
+
+class _BooleanEntry(_BaseEntry):
+    """Entry class for a boolean value.
+    
+    Attributes:
+        argInfo: Each key and its value are:
+            default: The default value.
+        checkBox: The checkbox showing the boolean value.
+    """
+
+    def __init__(self, name: str, argInfo: Dict[str, Any], parent: Optional[QWidget] = None):
+        """Extended."""
+        super().__init__(name, argInfo, parent=parent)
+        # widgets
+        self.checkBox = QCheckBox(self)
+        self.checkBox.setChecked(self.argInfo["default"])
+        # layout
+        self.layout.addWidget(self.checkBox)
+
+    def value(self) -> bool:
+        """Overridden.
+        
+        Returns the status of the checkBox.
+        """
+        return self.checkBox.isChecked()
+
+
+class _EnumerationEntry(_BaseEntry):
+    """Entry class for an enumeration value.
+    
+    Attributes:
+        argInfo: Each key and its value are:
+            choices: The pre-defined candidates.
+            default: The default value. If it does not exist, it is set to the first candidate.
+        comboBox: The combobox showing the enumeration value.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        argInfo: Dict[str, Any],
+        parent: Optional[QWidget] = None
+    ):
+        """Extended."""
+        super().__init__(name, argInfo, parent=parent)
+        # widgets
+        self.comboBox = QComboBox(self)
+        for choice in self.argInfo["choices"]:
+            self.comboBox.addItem(choice)
+        if self.argInfo["choices"]:
+            self.comboBox.setCurrentText(
+                self.argInfo["choices"][0] if self.argInfo["default"] is None
+                else self.argInfo["default"]
+            )
```

Comment:
I think you use `self.argInfo["choices"]` a lot.
If you want to, you can define a variable for it:
```python
choices = self.argInfo["choices"]
```
Or, moreover
```python
choices, default = map(self.argInfo.get, ("choices", "default"))
```

Of course, this is not necessary and you can keep your current code!

Reply:
Oh, it looks simpler! I will apply your method😃